File: missions/mission1.pas
program Mission1;

{$mode objfpc}

uses
  SysUtils, StrUtils;

function getinput : TStringArray;
const
  C_FNAME = '../inputs/1.txt';
var
  tfIn: TextFile;
  line_no : Integer = 0;
begin
  AssignFile(tfIn, C_FNAME);
  result := TStringArray.Create;
  SetLength(result, 1);
  try
    reset(tfIn);
    while not eof(tfIn) do
    begin
      if Length(result) < line_no + 1 then
        SetLength(result, Length(result) * 2);
      Readln(tfIn, result[line_no]);
      line_no := line_no + 1;
    end;
    CloseFile(tfIn);
    SetLength(result, line_no);
  except
    on E: EInOutError do
     writeln('File handling error occurred. Details: ', E.Message);
  end;
end;

var
  disabled_count : Integer = 0;
  max_time : Integer = 0;
  max_disabled_count : Integer = 0;
  max_disabled_time : Integer = 0;
  i, reactor, time, disabled_time, time_diff : Integer;
  data : TStringArray = ('');
  parts : TStringArray;
  reactor_times : Array of Integer = ();

begin
  data := getinput();
  for i := 0 to Length(data) - 1 do
  begin
      parts := SplitString(data[i], ' ');
      time := StrToInt(copy(parts[0], 3));
      reactor := StrToInt(copy(parts[1], 2));

      if parts[2] = 'disabled' then
        begin
          if Length(reactor_times) < reactor then
            SetLength(reactor_times, reactor);
          reactor_times[reactor] := time;

          disabled_count := disabled_count + 1;
          disabled_time := time;
        end
      else if parts[2] = 'enabled' then
        begin
          time_diff := time - reactor_times[reactor];
	  if time_diff > max_time then
            max_time := time_diff;

          if disabled_count > max_disabled_count then
            begin
              max_disabled_count := disabled_count;
              max_disabled_time := time - disabled_time;
            end
          else if disabled_count = max_disabled_count then
            begin
              time_diff := time - disabled_time;
              if time_diff > max_disabled_time then
                max_disabled_time := time_diff
            end;

          disabled_count := disabled_count - 1;
        end;
    end;

  writeln(max_time);
  writeln(max_disabled_time);
end.

File: missions/mission1.pl
#!/usr/bin/perl
use strict;
use warnings;
use List::Util qw(max);

my $fh;
open $fh, "<", "../inputs/1.txt";
my @data = <$fh>;
close $fh;
chomp($_) foreach @data;

my %times;
my $disabled_count = 0;
my $disabled_time;
my $max_time = 0;
my $max_disabled_count = 0;
my $max_disabled_time = 0;
for my $line (@data) {
	my ($time, $id, $state) = split/ /, $line;
	$time = substr($time, 2);
	$id = substr($id, 1);

	if ($state eq 'disabled') {
		$times{$id} = $time;
		$disabled_count++;
		$disabled_time = $time;
	} elsif ($state eq 'enabled') {
		$max_time = max($max_time, $time - $times{$id});

		if ($disabled_count > $max_disabled_count) {
              		$max_disabled_count = $disabled_count;
              		$max_disabled_time = $time - $disabled_time;
		} elsif ($disabled_count == $max_disabled_count) {
              		my $time_diff = $time - $disabled_time;
              		if ($time_diff > $max_disabled_time) {
                		$max_disabled_time = $time_diff;
			}
		}
          	
		$disabled_count--;
	}
}

printf("%d\n%d\n", $max_time, $max_disabled_time);
